frontend/src/components/widgets/DataFrame/EditingState.ts

```diff
@@ -0,0 +1,236 @@
+/**
+ * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { GridCell } from "@glideapps/glide-data-grid"
+
+import { notNullOrUndefined, isNullOrUndefined } from "src/lib/utils"
+
+import { BaseColumn } from "./columns"
+
+/**
+ * The editing state keeps track of all table edits applied by the user.
+ */
+class EditingState {
+  // row -> column -> GridCell
+  // Using [number, number] as a key for a Map would not work.
+  private editedCells: Map<number, Map<number, GridCell>> = new Map()
+
+  // List of rows represented by of column -> GridCell mappings
+  private addedRows: Array<Map<number, GridCell>> = []
+
+  // List of deleted row IDs
+  private deletedRows: number[] = []
+
+  // The original number of rows in the table (without potential additions & deletions)
+  private numRows = 0
+
+  constructor(numRows: number) {
+    this.numRows = numRows
+  }
+
+  /**
+   * Convert the current editing state to a JSON string.
+   *
+   * @param columns - The columns of the table
+   * @returns JSON string
+   */
+  toJson(columns: BaseColumn[]): string {
+    const columnsByIndex = new Map<number, BaseColumn>()
+    columns.forEach(column => {
+      columnsByIndex.set(column.indexNumber, column)
+    })
+
+    const currentState = {
+      edited_cells: {} as Record<string, any>,
```

Comment:
Yep, indeed. This is snake case because it's the format that is used in the widget state in the backend. Eslint did not complain.. does not seem to check those naming conventions.

Reply:
I added a comment to make that clear